File: python/752. Open the Lock.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        visited = set(deadends)
        if "0000" in visited:
            return -1
        if target == "0000":
            return 0
        q = ["0000"]
        answer = 0
        while q:
            answer += 1
            print(answer)
            size = len(q)
            for _ in range(size):
                curr = q.pop(0)
                if curr == target:
                    return answer
                if curr in visited:
                    continue
                visited.add(curr)
                for next in self.getNext(curr):
                    if next in visited:
                        continue
                    if next == target:
                        return answer
                    q.append(next)
        return -1
    def getNext(self, curr):
        result = []
        for i in range(4):
            for j in [-1, 1]:
                next = list(curr)
                next[i] = str((int(next[i]) + j) % 10)
                result.append("".join(next))
        return result

File: go/752. Open the Lock.go
func openLock(deadends []string, target string) int {
	visited := make(map[string]bool)
	for _, deadend := range deadends {
		visited[deadend] = true
	}
	if _, ok := visited["0000"]; ok {
		return -1
	}
	if target == "0000" {
		return 0
	}
	answer := 0
	q := []string{"0000"}
	for len(q) > 0 {
		size := len(q)
		for i := 0; i < size; i++ {
			current := q[0]
			q = q[1:]
			if current == target {
				return answer
			}
			if _, ok := visited[current]; ok {
				continue
			}
			visited[current] = true
			for _, next := range getNext(current) {
				if _, ok := visited[next]; !ok {
					q = append(q, next)
				}
			}
		}
		answer += 1
	}
	return -1
}

func getNext(current string) []string {
	next := []string{}
	for i := range current {
		next = append(next, current[:i]+string((int(current[i]-'0'+1)%10)+'0')+current[i+1:])
		next = append(next, current[:i]+string((int(current[i]-'0'+9)%10)+'0')+current[i+1:])
	}
	return next
}